### PR TITLE
accounts_index: Add SPL Token account indexing for token-2022 accounts

### PR DESCRIPTION
**Pull request by mvines**
token-2022 is now indexed by RPC nodes using the same mechanism as token.

### CODE REVIEW

**Review by joncinque — COMMENTED** [review]
I'm not in a position to approve, but good idea using bytemuck!

**Comment by mvines** [comment]
@CriesofCarrots - Do you know of a nice account index test that I could modify to get some coverage on the token-2022 additions in this PR?

**Comment by CriesofCarrots** [comment]
> Do you know of a nice account index test that I could modify to get some coverage on the token-2022 additions in this PR?

Adding some `inline_spl_token_2022::id()` cases [here](https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L4006) and/or [here](https://github.com/solana-labs/solana/blob/ac13d14e308b26b2d7b470ddcee02eed4e9e5167/runtime/src/accounts_index.rs#L4135) would probably be a good start

File: runtime/src/accounts_index.rs [hunk 1]
```diff
@@ -1602,26 +1602,42 @@ impl<T: IndexValue> AccountsIndex<T> {
         // 2) When the fetch from storage occurs, it will return AccountSharedData::Default
         // (as persisted tombstone for snapshots). This will then ultimately be
         // filtered out by post-scan filters, like in `get_filtered_spl_token_accounts_by_owner()`.
-        if *account_owner == inline_spl_token::id()
-            && account_data.len() == inline_spl_token::state::Account::get_packed_len()
-        {
+        if *account_owner == inline_spl_token::id() {
             if account_indexes.contains(&AccountIndex::SplTokenOwner) {
```
[the remark this answers is not in the document]
Comment (mvines) [reply runtime/src/accounts_index.rs:1606]: I looked at that and figured the pubkey equality check would generally be quicker or equal than a HashSet membership test, so didn't feel a strong need to change this around

File: runtime/src/inline_spl_token.rs [hunk 2]
```diff
@@ -21,11 +23,38 @@ pub const SPL_TOKEN_ACCOUNT_MINT_OFFSET: usize = 0;
...
+    }
+
+    pub fn unpack_account_owner(account_data: &[u8]) -> Option<&Pubkey> {
+        if account_data.len() != ACCOUNT_LENGTH {
+            None
+        } else {
+            Some(bytemuck::from_bytes(
+                &account_data
+                    [SPL_TOKEN_ACCOUNT_OWNER_OFFSET..SPL_TOKEN_ACCOUNT_OWNER_OFFSET + PUBKEY_BYTES],
+            ))
```
[the remark this answers is not in the document]
Comment (mvines) [reply runtime/src/inline_spl_token.rs]: All dryed out

**Review by CriesofCarrots — APPROVED** [review]
This turned out nice

**Pull request merged**